shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2020 Samuel Maddock <sam@samuelmaddock.com>.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/electron_api_web_frame_main.h"
+
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "base/lazy_instance.h"
+#include "base/logging.h"
+#include "content/browser/renderer_host/frame_tree_node.h"  // nogncheck
+#include "content/public/browser/render_frame_host.h"
+#include "gin/object_template_builder.h"
+#include "shell/browser/browser.h"
+#include "shell/common/gin_converters/frame_converter.h"
+#include "shell/common/gin_converters/gurl_converter.h"
+#include "shell/common/gin_converters/value_converter.h"
+#include "shell/common/gin_helper/dictionary.h"
+#include "shell/common/gin_helper/error_thrower.h"
+#include "shell/common/gin_helper/object_template_builder.h"
+#include "shell/common/gin_helper/promise.h"
+#include "shell/common/node_includes.h"
+
+namespace electron {
+
+namespace api {
+
+typedef std::unordered_map<content::RenderFrameHost*, WebFrame*> RenderFrameMap;
+base::LazyInstance<RenderFrameMap>::DestructorAtExit g_render_frame_map =
+    LAZY_INSTANCE_INITIALIZER;
+
+gin::WrapperInfo WebFrame::kWrapperInfo = {gin::kEmbedderNativeGin};
+
+// WebFrame can outlive its RenderFrameHost pointer so we need to check whether
+// its been disposed of prior to accessing it.
+#define CHECK_RENDER_FRAME(funcname, args, value)                      \
+  if (render_frame_disposed_) {                                        \
+    args->isolate()->ThrowException(v8::Exception::Error(              \
+        gin::StringToV8(args->isolate(),                               \
+                        "Render frame was disposed before "            \
+                        "WebFrame." #funcname " could be executed"))); \
+    return value;                                                      \
+  }
```

Comment:
Prefer a private method over a #define please. See e.g. `CheckAlive` in shell/browser/api/electron_api_tray.cc.